New Shoelace release 2.1.0

slhinyc opened the pull request:
### What's in the PR (Highlights)
- Updated `sl-switch` to include new layout options (label left, label left justified), plus some minor styling updates
- Updated `sl-radio` and `sl-radio-group` components:
  - Added form component options:`label-tooltip` and `context-note`
  - Added `horizontal` layout option
  - Updated how `contained` option is applied, so that it can be set just once from `sl-radio-group` and cascade to child `sl-radio` elements
  - Minor styling updates
- Added new `sl-checkbox-group` component, based on the existing `sl-radio-group` component:
  - So that a group of checkboxes can have a label, help text, label tooltip, and context note, just like other form components
  - Added a `horizontal` layout option to the Checkbox Group
  - `contained` option can be applied to the Checkbox Group, to match the way this option works for the Radio Group
- Small example updates and styling fixes to `sl-dialog`
- Small example updates to `sl-alert`, plus change default `sl-toast-stack` position from top right to bottom right
- Update `checked` and `x-log` system icons to use Font Awesome svgs

> 👉 Also see the [Changelog page](https://shoelace-bo92i7m0k-teamshares.vercel.app/teamshares/changelog) in the review app for a full list of changes for this release, which also includes changes from a previous PR that have already been merged to `next` but not yet tagged for release (we were holding off on cutting a new release until all the form-related components updates had been merged)
---
### ‼️ Note to reviewers: 
- Please review the new `sl-checkbox-group` component, including the tests and validity checks, closely. Although everything is working as expected, I'm not confident that I have done everything correctly here. I used the existing `sl-radio-group` as a model, but the behavior of the `sl-checkbox-group` is a bit different (multiple items can be checked; the group is valid if at least one checkbox in the group is checked), so I couldn't follow the `sl-radio-group` pattern exactly.
- There are a lot of other changes here, but most of them are styling and doc site example updates, so I'm less worried about those.
- There are also some new form component properties here -- the `label-tooltip`, `context-note` -- but these are following patterns already applied to other components & reviewed and approved by Daross in a previous PR that's already been merged to `next`

kdonovan commented on src/components/checkbox-group/checkbox-group.test.ts:
```diff
@@ -0,0 +1,298 @@
+      expect(el.checkValidity()).to.be.true;
+    });
+
+    it(`should be invalid when required and no checkboxes are checked`, async () => {
+      const el = await fixture<SlCheckboxGroup>(html`
+        <sl-checkbox-group label="Select an option" required>
+          <sl-checkbox name="option" value="1">Option 1</sl-checkbox>
+          <sl-checkbox name="option" value="2">Option 2</sl-checkbox>
+          <sl-checkbox name="option" value="3">Option 3</sl-checkbox>
+        </sl-checkbox-group>
+      `);
+
+      expect(el.checkValidity()).to.be.false;
+    });
```
Is this different than the check on line 12?

slhinyc replied:
I see what you mean. I was trying to follow the test patterns for the radio-group for these (these are the un-edited tests from the version of Shoelace we're using), and the `radio-group.test.ts` essentially has the same thing: 
- If you compare the tests at **line 13** vs **line 50** in this file: [src/components/radio-group/radio-group.test.ts](src/components/radio-group/radio-group.test.ts)

kdonovan commented on src/components/checkbox-group/checkbox-group.test.ts:
```diff
@@ -0,0 +1,298 @@
+    `);
+    const checkbox = checkboxGroup.querySelector<SlCheckbox>('#checkbox-1')!;
+    setTimeout(() => checkbox.click());
+    const event = (await oneEvent(checkboxGroup, 'sl-change')) as SlChangeEvent;
```
I'm missing where we're testing `sl-input` in this one

slhinyc replied:
This is another one where I was following the pattern in the same test for [src/components/radio-group/radio-group.test.ts](src/components/radio-group/radio-group.test.ts), but maybe it's a mistake in the Radio Group test? I just updated both to include an event listener for `sl-input`. Pushed a new commit!

kdonovan commented on src/components/checkbox-group/checkbox-group.component.ts:
```diff
@@ -0,0 +1,345 @@
+    if (this.value === null || this.value === undefined) {
+      this.initializeValueFromCheckboxes();
+    }
+    const anyCheckboxChecked = this.value.some(value => value.includes('true'));
+    const isRequiredAndEmpty = this.required && !anyCheckboxChecked;
+    const hasCustomValidityMessage = this.customValidityMessage !== '';
```
These are duplicated at least once below -- would they be better as helper methods rather than locals?

slhinyc replied:
This is another instance where I was following what's in the Radio Group component (the Shoelace default code), where some locals are repeated. Is it okay to leave, to keep things consistent across components?
- See **lines 103-104** and **lines 117-118** in this file: [src/components/radio-group/radio-group.test.ts](src/components/radio-group/radio-group.test.ts)

kdonovan commented on src/components/checkbox-group/checkbox-group.component.ts:
```diff
@@ -0,0 +1,345 @@
+
+  /** Gets the validity state object */
+  get validity() {
+    if (this.value === null || this.value === undefined) {
+      this.initializeValueFromCheckboxes();
+    }
```
Just my unfamiliarity here -- do we need this still despite it being called from `connectedCallback`?

slhinyc replied:
I added this because while all the tests were passing, I was getting this TypeError (see SS), and adding this `if` statement made it go away! I tried to adjust the tests a bunch to figure out why this is happening, but couldn't figure it out. If it's okay that I have this TypeError as long as the tests themselves are passing, then I can just delete this from the main component file, as everything is working as expected without this being repeated in the validity methods.
<img width="1130" alt="Screenshot 2024-05-06 at 16 05 44" src="https://github.com/teamshares/shoelace/assets/203228/f011bc87-3e77-4945-b60f-05022d2907b1">

kdonovan reviewed (changes requested):
Thanks for doing this, @slhinyc! Giant caveat that I'm super unfamiliar with all of this, but I got it spun up locally and everything looks generally good!

One likely issue: when I submit [the required checkbox group](http://localhost:4001/components/checkbox-group/) without selecting any, the `submit` event still fires -- and _then_ the browser-default validation error appears (sidenote: I don't know if it's under our control, but that message is a little confusing since it's actually submit _any_ of those boxes).

(I thought we had a test for this, but on second look I think we're only testing that `submit` isn't fired if we've already set `setCustomValidity`)

<img width="663" alt="Screenshot 2024-05-01 at 4 35 32 PM" src="https://github.com/teamshares/shoelace/assets/86022/cea06558-070f-4d93-867e-68e0586ad7a0">

<img width="470" alt="Screenshot 2024-05-01 at 4 35 44 PM" src="https://github.com/teamshares/shoelace/assets/86022/598a6f0d-b877-48b4-b0df-3e93d627290e">

slhinyc commented:
> One likely issue: when I submit [the required checkbox group](http://localhost:4001/components/checkbox-group/) without selecting any, the submit event still fires -- and then the browser-default validation error appears (sidenote: I don't know if it's under our control, but that message is a little confusing since it's actually submit any of those boxes).

So this looks like an error in the example code, which I copied from the Radio Group example, but it was wrong there, too! I figured out what was missing. Following the examples on the Form Controls doc page, we needed an `await Promise.all` to "Wait for controls to be defined before attaching form listeners". This seems to fix the issue & the examples now all work as expected. I also fixed the issue on two Radio Group validation examples, and also added a "Set custom validity" example to the Checkbox Group, as I realized that I was missing that one. 

This addresses all your comments, I believe? Let me know what you think!

kdonovan reviewed (approved):
checkboxgroup validation is working now -- LGTM! 🚗
